pgml-docs/docs/blog/scaling-postgresml-to-one-million-requests-per-second.md

```diff
@@ -0,0 +1,252 @@
+
+
+Our architecture has four components that may need to scale up or down based on load:
+
+1. Clients
+2. Load balancer
+3. [PgCat](https://github.com/levkk/pgcat/) pooler
+4. Postgres replicas
+
+We intentionally don't discuss scaling the primary in this post, because sharding, which is the most effective way to do so, is a fascinating subject that deserves its own series of posts. Spoiler alert: we sharded Postgres without any problems.
+
+#### Clients
+
+Clients are regular Postgres connections coming from web apps, job queues, or pretty much anywhere that needs data. They can be long-living or ephemeral and they typically grow in number as the application scales.
+
+Most modern deployments use containers which are added as load on the app increases, and removed as the load decreases. This is called dynamic horizontal scaling, and it's an effective way to adapt to changing traffic patterns experienced by most businesses.
+
+#### Load Balancer
+
+The load balancer is a way to spread traffic across horizontally scalable components, by routing new connections to targets in a round robin (or random) fashion. It's typically a very large box (or a fast router), but even those need to be scaled if traffic suddenly increases. Since we're running our system on AWS, this is already taken care of, for a reasonably small fee, by using an Elastic Load Balancer.
+
+#### PgCat
+
+<center>
+	<img src="https://raw.githubusercontent.com/levkk/pgcat/main/pgcat3.png" alt="PgCat" height="300" width="auto" /> <br />
+	_Meow. All your Postgres belong to me._
+</center>
+
+If you've used Postgres in the past, you know that it can't handle many concurrent connections. For large deployments, it's necessary to run something we call a pooler. A pooler routes thousands of clients to only a few dozen server connections by time-sharing when a client can use a server. Because most queries are very quick, this is a very effective way to run Postgres at scale.
+
+There are many poolers available presently, the most notable being PgBouncer, which has been around for a very long time, and is trusted by many large organizations. Unfortunately, it hasn't evolved much with the growing needs of highly available Postgres deployments, so we wrote [our own](https://github.com/levkk/pgcat/) which added important functionality we needed:
+
+- Load balancing of read queries
+- Failover in case a read replica is broken
+- Sharding (this feature is still being developed)
+
+In this benchmark, we used its load balancing feature to evenly distribute our XGBoost predictions across our Postgres replicas.
+
+
+#### Postgres Replicas
+
+Scaling Postgres reads is a solved problem. If more read queries are coming in, add a replica to serve the increased load. If the load is decreasing, remove a replica to save money. The data is replicated from the primary, so all replicas are identical, and all of them can serve any query, or in our case, an XGBoost prediction.
+
+
+## Results
```

Comment:
I would consider moving Results above Architecture, since it's what I immediately care about. IF the results are interesting, I'll be more invested in understanding the architecture.